include/elements/2d/quadrilateral_bspline_element.h

```diff
@@ -0,0 +1,245 @@
+      const VectorDim& xi, const Eigen::MatrixXd& nodal_coordinates,
+      const VectorDim& particle_size,
+      const VectorDim& deformation_gradient) const override;
+
+  //! Return the type of shape function
+  mpm::ShapefnType shapefn_type() const override {
+    return mpm::ShapefnType::BSPLINE;
+  }
+
+  //! Return number of shape functions
+  unsigned nfunctions() const override { return nconnectivity_; }
+
+  //! Return if natural coordinates can be evaluates
+  bool isvalid_natural_coordinates_analytical() const override { return false; }
+
+  //! Compute Natural coordinates of a point (analytical)
+  //! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+  //! \param[in] point Location of the point in cell
+  //! \retval xi Return the local coordinates
+  VectorDim natural_coordinates_analytical(
+      const VectorDim& point,
+      const Eigen::MatrixXd& nodal_coordinates) const override;
+
+  //! Assign nodal connectivity property for bspline elements
+  //! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+  //! \param[in] nodal_properties Vector determining node type for each
+  //! dimension
+  void initialise_bspline_connectivity_properties(
+      const Eigen::MatrixXd& nodal_coordinates,
+      const std::vector<std::vector<unsigned>>& nodal_properties) override;
+
+  //! Return the degree of shape function
+  mpm::ElementDegree degree() const override {
+    return mpm::ElementDegree::Quadratic;
+  };
+
+ private:
+  //! Compute B-Spline Basis Function using the recursive De Boor's algorithm
+  //! for single direction
+  //! \param[in] point_coord point coordinate in one direction
+  //! \param[in] nodal_coord nodal coordinate in one direction
+  //! dimension
+  //! \param[in] knot_vector Knot vector type associated with direction
+  //! \param[in] poly_order Polynomial degree
+  //! \param[in] index Index associated to local knot_vector
+  double kernel(double point_coord, double nodal_coord, unsigned node_type,
+                unsigned poly_order, unsigned index = 0) const;
+
+  //! Compute B-Spline Basis Function Gradient using the recursive De Boor's
+  //! algorithm for single direction
+  //! \param[in] point_coord point coordinate in one direction
+  //! \param[in] nodal_coord nodal coordinate in one direction dimension
+  //! \param[in] knot_vector Knot vector type associated with direction
+  //! \param[in] poly_order Polynomial degree
+  //! \param[in] index Index associated to local knot_vector
+  double gradient(double point_coord, double nodal_coord, unsigned node_type,
+                  unsigned poly_order, unsigned index = 0) const;
+
+  //! Function that returns BSpline knot vector
+  //! The order of the vectors are:
+  //! Regular = 0,
+  //! LowerBoundary = 1,
+  //! LowerIntermediate = 2,
+  //! UpperIntermediate = 3
+  //! UpperBoundary = 4,
+  //! LowerBoundary2 = 5,
+  //! UpperBoundary2 = 6
+  std::vector<double> knot(unsigned node_type) const {
```

Comment:
What are the LowerBoundary2 and UpperBoundary2? Is there any clearer naming?

Reply:
Good questions! The `LowerBoundary2` and `UpperBoundary2` are actually virtual and added only to satisfy the partition of unity at `LowerBoundary` and `UpperBoundary` types, respectively! So the user only needs to determine the `LowerBoundary` and `UpperBoundary` node sets, but not the virtual one.